compiler/qsc_qasm3/src/compile.rs

```diff
@@ -0,0 +1,4364 @@
+                            let kind = SemanticErrorKind::PowModifierMustHaveExponent(span);
+                            self.push_semantic_error(kind);
+                            return GateModifier::Pow(None, span);
+                        };
+                        extract_pow_exponent(&expr, span)
+                    }
+                    Modifier::CtrlModifier(ctrl_mod) => {
+                        let ctrls = self.extract_controls_from_modifier(ctrl_mod.paren_expr());
+                        GateModifier::Ctrl(ctrls, span)
+                    }
+                    Modifier::NegCtrlModifier(neg_ctrl_mod) => {
+                        let ctrls = self.extract_controls_from_modifier(neg_ctrl_mod.paren_expr());
+                        GateModifier::NegCtrl(ctrls, span)
+                    }
+                }
+            })
+            .collect::<Vec<_>>();
+
+        if let Some(gate_call_expr) = modified_gate_call_expr.gate_call_expr() {
+            self.compile_gate_call_expr_impl(&gate_call_expr, expr_span, &modifiers)
+        } else {
+            let Some(gphase_call_expr) = modified_gate_call_expr.g_phase_call_expr() else {
+                // error
+                return None;
+            };
+            self.compile_gphase_call_expr_impl(&gphase_call_expr, expr_span, &modifiers)
+        }
+    }
+
+    /// Extracts the literal int from `ctrl(value)` and `negctrl(value)` modifiers.
+    /// Returns None if the modifier is invalid and pushes an error.
+    /// Returns Some(1) if the modifier is empty.
+    /// Returns Some(value) if the modifier is valid.
+    fn extract_controls_from_modifier(&mut self, paren_expr: Option<ParenExpr>) -> Option<usize> {
+        if let Some(paren_expr) = paren_expr {
+            if let Some((ctrl, sign)) = compile_paren_lit_int_expr(&paren_expr) {
+                if sign {
+                    let kind = SemanticErrorKind::NegativeControlCount(span_for_syntax_node(
+                        paren_expr.syntax(),
+                    ));
+                    self.push_semantic_error(kind);
+                }
+                Some(ctrl)
+            } else {
+                let kind = SemanticErrorKind::InvalidControlCount(span_for_syntax_node(
+                    paren_expr.syntax(),
+                ));
+                self.push_semantic_error(kind);
+                None
+            }
+        } else {
+            Some(1)
+        }
+    }
+
+    fn compile_paren_expr(
+        &mut self,
+        paren_expr: &oq3_syntax::ast::ParenExpr,
+    ) -> Option<QasmTypedExpr> {
+        let span = span_for_syntax_node(paren_expr.syntax());
+        let expr = paren_expr.expr()?;
+        let texpr = self.compile_expr(&expr)?;
+        let pexpr = ast_builder::wrap_expr_in_parens(texpr.expr, span);
+        Some(QasmTypedExpr {
+            ty: texpr.ty.clone(),
+            expr: pexpr,
+        })
+    }
+
+    #[allow(clippy::too_many_lines)]
```

Comment:
Do we need this `allow` here? This function is about 50 lines long, and the `too_many_lines` lint triggers when you have over 100 lines of code.